app/resources/services/integrations/coil.js

```diff
@@ -0,0 +1,167 @@
+const config = require('config')
+const logger = require('../../../../lib/logger.js')()
+const btoa = require('btoa')
+const { User } = require('../../../db/models')
+const passport = require('passport')
+const OAuth2Strategy = require('passport-oauth').OAuth2Strategy
+const axios = require('axios')
+const InternalOAuthError = require('passport-oauth2').InternalOAuthError
+const { findUser } = require('./findUser')
+
+const authToken = btoa(
+  process.env.COIL_CLIENT_ID +
+    ':' +
+    encodeURIComponent(process.env.COIL_CLIENT_SECRET)
+)
+const coilStrategy = new OAuth2Strategy(
+  {
+    authorizationURL: 'https://coil.com/oauth/auth',
+    tokenURL: 'https://coil.com/oauth/token',
+    clientID: process.env.COIL_CLIENT_ID,
+    clientSecret: process.env.COIL_CLIENT_SECRET,
+    scope: 'simple_wm openid',
+    callbackURL: `${config.restapi.protocol}${config.app_host_port}/services/integrations/coil/callback`,
+    passReqToCallback: true,
+    customHeaders: {
+      authorization: `Basic ${authToken}`,
+      content_type: 'application/x-www-form-urlencoded'
+    }
+  },
+  async function (req, accessToken, refreshToken, params, profile, done) {
+    // params are returned by passport from the request
+    const databaseUser = await findUser(req.query.state)
+    // passing the profile data along the request, probably another way to do this
+    // we get the access token here, which we may need to then pass to the next endpoint for coil to get the user info (subscriptions, etc)
+    req.profile = profile
+    profile.refresh_token = refreshToken
+    return done(null, databaseUser)
+  }
+)
+
+const getUserInfo = async (accessToken, done) => {
+  try {
+    const requestConfig = {
+      method: 'post',
+      url: 'https://api.coil.com/user/info',
+      headers: {
+        Authorization: 'Bearer ' + accessToken,
+        content_type: 'application/x-www-form-urlencoded'
```

Comment:
Does this way of specifying a header work? Typically, HTTP headers use dashes instead of underscores, e.g.:

```js
headers: {
   Authorization: 'Bearer ' + accessToken,
   'content-type': 'application/x-www-form-urlencoded'
}
```

This does mean you need to quote the property name, and it will look mismatched with other property names that do not need to be quoted, but this is closer to the expected implementation. When in doubt, I'd refer to their usage in their own documentation:
https://github.com/axios/axios#using-applicationx-www-form-urlencoded-format

Reply:
the requests have been working, but I agree about following convention